src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -14,35 +14,288 @@ namespace Microsoft.CodeAnalysis
     /// </summary>
     public class ProjectDependencyGraph
     {
-        private readonly ImmutableArray<ProjectId> _projectIds;
+        private readonly ImmutableHashSet<ProjectId> _projectIds;
         private readonly ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> _referencesMap;
 
         // guards lazy computed data
         private readonly NonReentrantLock _dataLock = new NonReentrantLock();
 
-        // these are computed fully on demand
+        // These are computed fully on demand. null or ImmutableArray.IsDefault indicates the item needs to be realized
         private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> _lazyReverseReferencesMap;
```

Comment:
@CyrusNajmabadi Not entirely understanding how that's different than what's implemented in this PR...?

Reply:
I dunno.  your appraoch (currently have only skimmed it) seems much more complex.  Mine would literally just be something like:

```c#
var projectsToUpdate = GetStackFromPool();
projectsToUpdate.Add(projectId);

while (projectsToUpdate.Any())
{
    var currentProject = projectToUpdate.Pop();
    transitiveReferences[currentProject] = transitiveReferences[currentProject].Union(referencedProjectIds);

  projetsToUpdate.PushRange(_directReverseReferences[projectId]);
}

ReturnStackToPool();
```

That's it...

Reply:
Basically: ComputeNewTransitiveReferencesMapForAdditionalProjectReferences seems much more complex than i would expect.  Given direct forward/backward mappings, it feels like it should be a pretty trivial function to write...

Reply:
> it feels like it should be a pretty trivial function to write...

That's what I thought on Friday. I got it wrong.

Much of the complexity of mine is coming from the fact that we don't always have all the transitive map is often incomplete: prior to my change, we compute that project-at-a-time. In your pseudocode, you would need additional checking that transitiveReferences might not have that entry, and if so then you have to preserve that. I could alternately choose to investigate that further, but at this point I feel we're entering bikeshedding territory.

Your code is also missing one important point: you're only propagating the new direct references. If A -> B and C -> D, and I'm adding B -> C, then you need to get D added too. My first loop computes you need to include D, also accounting for the fact we might not know C's transitive references and thus instead need to propagate "we don't know".

Reply:
> Your code is also missing one important point: you're only propagating the new direct references. If A -> B and C -> D, and I'm adding B -> C, then you need to get D added too. My first loop computes you need to include D, also accounting for the fact we might not know C's transitive references and thus instead need to propagate "we don't know".

Ack.  That feeling when you realize you're an idiot :D

Reply:
> Much of the complexity of mine is coming from the fact that we don't always have all the transitive map is often incomplete: prior to my change, we compute that project-at-a-time. In your pseudocode, you would need additional checking that transitiveReferences might not have that entry, and if so then you have to preserve that. I could alternately choose to investigate that further, but at this point I feel we're entering bikeshedding territory.

No.  that's fair.  I just wanted to ask up front before investinga lot of time on this PR in those two functions.  Note: i will only be able to get to htem later today.  They're complex enough that i want to be able to really think through them to convince myself they're ok.  

Thanks for teh good explanations :)

Reply:
Other possible perf benefit for mine is there are strictly O(n) Union operations happening (at most one per updated project), whereas yours is O(m) where m is number of project references which might often be O(n^2). I was careful to use builders whenever sanely possible to reduce allocations.

Reply:
> Ack. That feeling when you realize you're an idiot :D

That's how I felt on Friday. In my case, I made the reverse mistake: I was correctly adding both C and D to B's transitive references, but forgot to update A too.

Reply:
@CyrusNajmabadi You're also welcome to not review this and let somebody else on the team do it if you don't have time. I recall these are algorithms you tend to like -- and at least I nerd-sniped you!

Reply:
I would def get multiple eyes on this if i were you :)   I was just letting you knwo that it will take me 24 hours, and that my lakc of further comments wasn't me being done, just me not having the time *right now* :)